lib/util/Ordering.ts

```diff
@@ -79,49 +77,42 @@ export function orderTypes(termA: RDF.Term | undefined, termB: RDF.Term | undefi
-
-  const openWorldType: ISuperTypeProvider = {
-    discoverer: typeDiscoveryCallback || (() => 'term'),
-    cache: typeCache || new LRUCache(),
+  const isLess = regularFunctions[C.RegularOperator.LT];
+  const isEqual = regularFunctions[C.RegularOperator.EQUAL];
+  const context = {
+    now: new Date(),
+    functionArgumentsCache: {},
+    superTypeProvider: {
+      discoverer: typeDiscoveryCallback || (() => 'term'),
+      cache: typeCache || new LRUCache(),
+    },
+    defaultTimeZone: { zoneHours: 0, zoneMinutes: 0 },
   };
-  const termTransformer = new TermTransformer(openWorldType);
+
+  const termTransformer = new TermTransformer(context.superTypeProvider);
   const myLitA = termTransformer.transformLiteral(litA);
   const myLitB = termTransformer.transformLiteral(litB);
-  const typeA = myLitA.dataType;
-  const typeB = myLitB.dataType;
-
-  const superTypeDictA: GeneralSuperTypeDict = getSuperTypeDict(typeA, openWorldType);
-  const superTypeDictB: GeneralSuperTypeDict = getSuperTypeDict(typeB, openWorldType);
 
-  // Special handling of specific datatypes
-  if (!isNonLexicalLiteral(myLitA) && !isNonLexicalLiteral(myLitB)) {
-    if (TypeURL.XSD_BOOLEAN in superTypeDictA && TypeURL.XSD_BOOLEAN in superTypeDictB ||
-      TypeAlias.SPARQL_NUMERIC in superTypeDictA && TypeAlias.SPARQL_NUMERIC in superTypeDictB ||
-      TypeURL.XSD_STRING in superTypeDictA && TypeURL.XSD_STRING in superTypeDictB) {
-      return comparePrimitives(myLitA.typedValue, myLitB.typedValue);
+  try {
+    // TODO: Calling apply does come at a small cost, searching for the correct overload...
+    //  @Ruben, we can fix this in an ugly way by using if statements (like it was before).
+    //  If we than also add an option to this function that will disable the fallback,
+    //  we could use the this function in our regular functions.
+    if ((<E.BooleanLiteral> isEqual.apply([ myLitA, myLitB ], context)).typedValue) {
+      return 0;
     }
-    if (TypeURL.XSD_DATE_TIME in superTypeDictA && TypeURL.XSD_DATE_TIME in superTypeDictB) {
-      return comparePrimitives(
-        toUTCDate(myLitA.typedValue, defaultTimezone).getTime(),
-        toUTCDate(myLitB.typedValue, defaultTimezone).getTime(),
-      );
+    if ((<E.BooleanLiteral> isLess.apply([ myLitA, myLitB ], context)).typedValue) {
+      return -1;
     }
-    if (TypeURL.RDF_LANG_STRING in superTypeDictA && TypeURL.RDF_LANG_STRING in superTypeDictB) {
-      const compareType = comparePrimitives(myLitA.typedValue, myLitB.typedValue);
-      if (compareType !== 0) {
-        return compareType;
-      }
-      return comparePrimitives((<LangStringLiteral>myLitA).language, (<LangStringLiteral>myLitB).language);
+    return 1;
+  } catch {
+    // Fallback to string-based comparison
+    const compareType = comparePrimitives(myLitA.dataType, myLitB.dataType);
+    if (compareType !== 0) {
+      return compareType;
     }
+    return comparePrimitives(myLitA.str(), myLitB.str());
```

Comment:
If we add an extra parameter to the function that makes the catch rethrow instead of executing the fallback, that could allow us to easily fix https://github.com/comunica/sparqlee/issues/178

Reply:
The same `strict` parameter as the function above you mean?